### PR TITLE
Feature/330 customize blueprint

### LINKED ISSUE #330: Blueprint name collision
**Describe the bug**
Hello,

I'm getting a blueprint name collision error.
```
AssertionError: A blueprint's name collision occurred between <flask.blueprints.Blueprint object at 0x0000021E18B91E48> and <flask.blueprints.Blueprint object at 0x0000021E16FA4940>.  Both share the same name "dashboard".  Blueprints that are created on the fly need unique names.
```
That's true because I have a route called "dashboard" already, but I changed the custom route link as described in the documentation as:

```
[dashboard]
APP_VERSION=1.0
CUSTOM_LINK=monitoring
MONITOR_LEVEL=3
OUTLIER_DETECTION_CONSTANT=2.5
SAMPLING_PERIOD=20
ENABLE_LOGGING=True

[authentication]
USERNAME=admin
PASSWORD=admin
GUEST_USERNAME=guest
GUEST_PASSWORD=['dashboardguest!', 'second_pw!']
SECURITY_TOKEN=cc83733cb0af8b884ff6577086b87909

[visualization]
TIMEZONE=Europe/Amsterdam
COLORS={'main':'[0,97,255]',
        'static':'[255,153,0]'}
```

my init looks as follows:

```
import flask_monitoringdashboard as monitoring_dashboard
app_service = Flask(__name__, )
monitoring_dashboard.config.init_from(file="../../monitoring_config.cfg")
monitoring_dashboard.bind(app_service)
app_service.config['JSON_SORT_KEYS'] = False
app_service.register_blueprint(routes.register.api, url_prefix='/register')
app_service.register_blueprint(routes.login.api, url_prefix='/login')
app_service.register_blueprint(routes.dashboard.api, url_prefix='/dashboard')
app_service.register_blueprint(routes.payment.api, url_prefix='/payment')
..
.
.
```

When I try to change the blueprint names myself like:

```
monitoring_dashboard.config.link = "monitoring"
monitoring_dashboard.blueprint.name = "monitoring"
```

it seems to be woring but when I call `localhost/monitoring/` I'm getting a:

`werkzeug.routing.BuildError: Could not build url for endpoint 'dashboard.index'. Did you mean 'dashboard.get_users_increase' instead?`

That's really weird, why does it not seem to work with the adjusted config file?

### PR DESCRIPTION
**Pull request by bogdanp05**
Support configurable FMD blueprint name via the env var FMD_BLUEPRINT. Fixes #330.
Btw, seeing how in the past 2 months there's been a lot of new stuff merged into develop, I decided to bump the version to 3.1.0 instead of 3.0.10.

### CODE REVIEW

File: config.cfg [hunk 1]
```diff
@@ -2,7 +2,7 @@
 APP_VERSION=1.0
 GIT=/<path to your project>/.git/
 CUSTOM_LINK=dashboard
-MONITOR_LEVEL=3
+MONITOR_LEVEL=1
```
Comment (mircealungu) [inline config.cfg:5]: good!

File: codecov.yml [hunk 2]
```diff
@@ -3,6 +3,4 @@ codecov:
 
 coverage:
   status:
-    patch:
-      default:
-        threshold: 1%
+    patch: off
```
Comment (mircealungu) [inline codecov.yml:6]: why are we disabling this?
does this not disable codecov for all the PRs? 
is this what you want?

File: flask_monitoringdashboard/__init__.py [hunk 3]
```diff
@@ -19,15 +19,14 @@
...
-
 
 def loc():
     """Get the current location of the project."""
     return os.path.abspath(os.path.dirname(__file__)) + '/'
 
 
-blueprint = Blueprint('dashboard', __name__, template_folder=loc() + 'templates')
+config = Config()
+blueprint = Blueprint(config.blueprint_name, __name__, template_folder=loc() + 'templates')
```
Comment (mircealungu) [inline flask_monitoringdashboard/__init__.py:29]: eheh. as the number of users grows, we'll end up extracting all the hardcoded strings into configuration options :)

File: flask_monitoringdashboard/core/config/__init__.py [hunk 4]
```diff
@@ -27,6 +27,7 @@ def __init__(self):
         """Sets the default values for the project."""
         # dashboard
         self.version = '1.0'
+        self.blueprint_name = os.environ.get('FMD_BLUEPRINT', 'dashboard')
```
Comment (mircealungu) [inline flask_monitoringdashboard/core/config/__init__.py:30]: sorry, why don't we read this from the config file?

File: codecov.yml (hunk 2 above)
Comment (bogdanp05) [reply codecov.yml:6]: The message was `@codecov codecov/patch — 70.00% of diff hit (target 91.36%)` with a failed status. According to the [docs](https://docs.codecov.io/docs/commit-status#patch-status), the 70% comes from the coverage of the diff and it was required to be at least as high as the project coverage of 91.36%, which is already very high. I don't know if we want to enforce such a strict coverage requirement on all PRs.

Comment (FlyingBird95) [reply codecov.yml:6]: I'm fine with removing the patch codecov 👍

File: flask_monitoringdashboard/core/config/__init__.py (hunk 4 above)
Comment (FlyingBird95) [reply flask_monitoringdashboard/core/config/__init__.py:30]: I agree with mircea that this should come from the config file

Comment (bogdanp05) [reply flask_monitoringdashboard/core/config/__init__.py:30]: The `init_from` method of the `Config` class is called by the user after the FMD is imported. However, the blueprint is created in FMD's `__init__` itself, so we need the blueprint name there. If we want to read it from a file at import time we need to either: 
a) fix the location of the config file itself 
b) fix the name of the env var to point to the location of the config file.
Now I'm doing option b)  but minus the extra step of pointing to a file, I'm giving the value directly.

File: codecov.yml (hunk 2 above)
Comment (mircealungu) [reply codecov.yml:6]: Aha! 
Thanks for the clarification Bogdan!

File: flask_monitoringdashboard/core/config/__init__.py (hunk 4 above)
Comment (mircealungu) [reply flask_monitoringdashboard/core/config/__init__.py:30]: ouch! i see now.

Comment (mircealungu) [reply flask_monitoringdashboard/core/config/__init__.py:30]: wouldn't it work if the blueprint would be initialized with None 
and then within the `.bind` that global blueprint would be overwritten with whatever's coming from config if there's something ?

Comment (bogdanp05) [reply flask_monitoringdashboard/core/config/__init__.py:30]: Good idea, it should work. I'll give it a try

Comment (bogdanp05) [reply flask_monitoringdashboard/core/config/__init__.py]: @mircealungu Turns out it's not a good idea to create a new blueprint object with every call of `bind`.  The first call of `bind` works fine, but then starting with the second call, the url's don't get registered anymore, hence the failing tests. Scratched my head for a while as to why, but in the end it was enough to just keep the global blueprint and only change its name in `bind`.

Comment (mircealungu) [reply flask_monitoringdashboard/core/config/__init__.py]: Changing the name sounds much more elegant! Nice :)

**Comment by saravanakesavan** [comment]
Hey @bogdanp05, @FlyingBird95 

Is this changes available in 3.1.0 release of Flask-MonitoringDashboard?

**Comment by saravanakesavan** [comment]
Hi @FlyingBird95 @mircealungu @bogdanp05 ,

I am trying to give a BLUEPRINT_NAME in config. While trying that the UI page is not loading. 

http://localhost:5001/xyzdashboard

I am getting the following error:

raise BuildError(endpoint, values, method, self)
werkzeug.routing.BuildError: Could not build url for endpoint 'xyzdashboard.index'. Did you mean 'dashboard.index' instead?

**Pull request merged**